### DISCUSSION TITLE
How to mix multiple guards and an "async" guard?

### DISCUSSION THREAD

**Discussion by maxarias-io** [statelyai/xstate#3197] [Q&A] [answered]
Hey, so this is more of a conceptual question I guess.

If I understood correctly, an "async guard" is an anti-pattern and that async guard should be represented as its own state machine with different targets based on the outcomes.

I have an event that has multiple guards and depending on the guard that passes, it goes to a different state. One of the potential guard needs to run async code.
The order in which these are executed is important to me, the preceding guard is more important than the next and so on.

My question is, how can I mix both of these "type" of guards so that they run in sequence. If that "async guard" fails, I want it to go to the next possible guard, not a different target.

Is it possible? Thanks!

**Comment by davidkpiano** [comment, marked as answer]
You can use invokes for "async guards" and transient states for "sync guards":

```js
const machine = createMachine({
  initial: 'idle',
  states: {
    idle: {
      on: { EVENT: 'asyncGuard'}
    },
    asyncGuard: {
      invoke: {
        src: 'someAsyncCheck',
        onDone: 'guard2',
        onError: 'failure'
      }
    },
    guard2: {
      always: [
        { cond: 'someSyncCheck', target: 'success' },
        { target: 'failure'}
      ]
    },
    success: {},
    failure: {}
  }
})
```
![CleanShot 2022-04-06 at 17 27 27](https://user-images.githubusercontent.com/1093738/162074256-3f6a5732-eae1-4eaa-be5c-f2cd9327d250.png)

**Reply by maxarias-io** [reply]
Ah I see, so I'll need to use different type of states instead of an array of guards.

I was doing something like:

```
{
  "id": "Machine",
  "initial": "IDLE",
  "states": {
    "IDLE": {
      "on": {
        "EVENT_ONE": [
          {
            "cond": "guard1",
            "target": "target1"
          },
          {
            "cond": "guard2",
            "target": "target2"
          },
          {
            "cond": "guard3",
            "target": "target3"
          }
        ]
      ...

```

I'll try it out, thanks for the quick response!
